### ISSUE TITLE
Ability to have a changelog

### ISSUE THREAD

**Issue by InTheCloudDan** [goreleaser/nfpm#42]
I was evaluating nfpm and it would be nice to have the ability to include a changelog file in the generated package. Right now I don't see any way to do it.

**Comment by djgilcrease** [comment]
Would you want this to show up as the rpm changelog or just be a file included in the package?

**Comment by djgilcrease** [comment]
https://www.debian.org/doc/debian-policy/ch-source.html#s-dpkgchangelog
http://packaging.ubuntu.com/html/debian-dir-overview.html#the-changelog
https://rpm-guide.readthedocs.io/en/latest/rpm-guide.html

Both RPM and Deb have specific formats required for changelogs that differ

**Debian Format**
```
<Package Name> (<Version>-<Release>+<Metadata>) <distribution> ; urgency=low

 * List (Closes: #nnnn)  <nnnn is the bug number of your ITP>
 * of
 * Changes

-- <Packager Name and Email>  <packged date Mon, 22 Mar 2010 00:37:31 +0100>
```

**RPM Format**
```
* <packged date Tue May 31 2016> <Packager Name and Email> - <Version>-<Release>
- List (Closes: #nnnn)
- of
- changes
```

**Data that we already have**
- Version
- Release
- deb-Metadata
- Maintainer
- packaged date


**Data We need**
- DEB: distribution
- DEB: urgency

We will require changelog schema that can be tranformed into both formats. I think the schema (or at least it's auto generation) should be owned by goreleaser as it will help with https://github.com/goreleaser/goreleaser/issues/1131 & https://github.com/goreleaser/goreleaser/issues/990 then nfpm can use that to create the correct changelog data for each format

**Comment by djgilcrease** [comment]
Idea for format that I think will allow people the most flexibility to generate changelog files in the format(s) they wish to support not just `.dep` & `.rpm`

```yaml
- semver: 1.0.0+git.e6c8682
  date: 2006-01-02T15:04:05.999999999Z07:00 // RFC3339Nano
  packager: Carlos A Becker <pkg@carlosbecker.com>
  notes:
    header: |
      New Features
      ============
      There are major new features in this release

      Known Issues
      ============
      bla bla bla
    footer: |
      Header section
      =============
      More things here

      Bla Bla Bla
      =============
      More things here
  changes:
    - commit: e6c8682
      note: Use Version instead of Git tag to for .deb/.rpm
      closes:
        - 1157
    - commit: 50235f2
      note: chore(deps): bump all 
      closes:
        - 1174
    - commit: eb1a122
      note: fix: lint issues
  deb:
    urgency: low
    distributions:
      - trusty
      - stable
- semver: 1.0.0-0.1.rc2+git.39d07c3
  date: 2006-01-02T15:04:05.999999999Z07:00 // RFC3339Nano
  packager: Carlos A Becker <pkg@carlosbecker.com>
  changes:
    - commit: 39d07c3
      note: feat: git diff when dirty
      closes:
        - 1178
  deb:
    urgency: low
    distributions:
      - trusty
      - stable
```

**Go Structs**
```go
type ChangeLogEntries []*ChangeLog

type ChangeLog struct {
    ChangeLogOverridables `yaml:",inline"`
    Semver string `yaml:"semver"`
    Date time.Time `yaml:"date"`
    Packager: string `yaml:"packager"`
    Notes *ChangeLogNotes `yaml:"notes,omitempty"`
    Changes ChangeLogChanges `yaml:"changes,omitempty"`
}

type ChangeLogOverridables struct {
    Deb          *ChangelogDeb               `yaml:"deb,omitempty"`
}

type ChangelogDeb struct {
    Urgency string `yaml:"urgency"`
    Distributions []string `yaml:"distributions"`
}

type ChangeLogNotes struct {
    Header *string `yaml:"header,omitempty"`
    Footer *string `yaml:"footer,omitempty"`
}

type ChangeLogChanges []*ChangeLogChange

type ChangeLogChange struct {
    Commit: string `yaml:"commit"`
    Note string `yaml:"note"`
    Closes []string `yaml:"closes,omitempty"`
}
```

@caarlos0 feedback / comments on this idea? This almost feels like a new project that  is a `cli / lib` so that it can be used by both `nfpm`, `goreleaser`, or independently

**Comment by caarlos0** [comment]
It looks good! I think most people will want to generate the changelog though... 

Extracting the `commit` and `title` from the commit log seems simple enough, but I'm not sure about the closes...

Would it work if instead we just passed a list of strings:

```yaml
# ...
  changes:
    - 39d07c3 feat: git diff when dirty (closes #123)
    - e6c8682 fix: something
# ...
```

If deb and rpm accept that, I think its simpler to implement...

**Comment by djgilcrease** [comment]
Would be yes. I will do up a quick `cli / lib` project in my name space to get the basic API down, then we can move it into the `goreleaser` namespace once you feel it is ready to use in both `goreleaser` and `nfpm`

**Comment by caarlos0** [comment]
awesome, thanks @djgilcrease !

**Comment by djgilcrease** [comment]
take a look @ https://github.com/djgilcrease/chglog next week I will do a branch of nfpm that makes use of the `chglog` to create the deb and rpm changelogs

**Comment by caarlos0** [comment]
will do!

btw, if you want to move it under goreleaser's org I'm totally fine with it, your call :)

**Comment by djgilcrease** [comment]
@eugene-babichenko can you check if the .deb package created by nfpm on the https://github.com/goreleaser/nfpm/tree/feature/use-chglog branch has what you need and if not can you detail the steps you are doing after the package is created so I can try to test it?

**Comment by djgilcrease** [comment]
merged #158

**Comment by caarlos0** [comment]
This issue has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
